File: roles/inventoryplay/files/inventory.py
#!/usr/bin/env python3
import os
import subprocess
import argparse
import pathlib
import re

debug = False

def read_all_file(dir):

    hash = {};
    if not dir:
        return hash

    try:
        isdir = os.path.isdir(dir)
    except Exception as e:
        print("Error checking status of a directory","error:",e)
        return hash

    if not isdir:
        return hash

    files = [];
    with os.scandir(dir) as it:
        for entry in it:
            if debug: print("entry:",entry)
            if not entry.is_file():
                continue

            if entry.name.startswith('.'):
                continue
           
            #basename = os.path.basename(entry.name)
            #files.append(basename)
            files.append(os.path.join(dir, entry.name))

    for file in files:
        array = [];
        if debug: print("file:",file)
        try:
             f = open(os.fsdecode(file), "r")
        except Exception as e:
            print("Error reading file", file, " error: ", e)
            continue
        else:
            for line in f:
                array.append(line.strip())

            f.close()
            hash[os.path.basename(file)] = array

    return hash

def get_all_playbooks(pb,hash):
    res = []

    if not pb:
        return res

    if debug: print("Searching for",pb,"in",hash)
    for target in hash:
        arr = hash[target]
        if len(arr):
            targetplay = arr[0]
            if debug: print("target:",target,"targetplay:",targetplay)
            if targetplay:
                if targetplay == pb:
                    if debug: print("found one")
                    res.append(target)
                    arr.pop(0)


    return res

def construct_runs(hash):
    playbook = []

    run = 0
    found = 1

    while found:
        found = 0
        run += 1

        for target in hash:
            if debug: print("target:",target)
            arr = hash[target]
            if not len(arr):
                continue

            pb = arr[0]

            if not pb:
                continue

            if debug: print("pb:",pb)

            arrlen = len(playbook)
            if debug: print("len:",arrlen,"run:",run)
            if arrlen < run:
                play = dict()
                play["playbook"]=pb
                play["targets"]= get_all_playbooks(pb, hash)

                playbook.append(play)
                found = 1

    return playbook


if __name__ == "__main__": 
    parser = argparse.ArgumentParser(description="inventoryplay")
    parser.add_argument('--directory','-d', required=True, type=pathlib.Path, help='playbook inventory')
    parser.add_argument('--inventory','-i', required=True, action='append', type=pathlib.Path, help='inventory')
    parser.add_argument('--debug', action='store_true', help='enable debug')
    args = parser.parse_args()
    directory = args.directory
    debug = args.debug
    inventory = args.inventory

    hash = read_all_file(directory)
    run = construct_runs(hash)
    for i in range(len(run)):
        print("run:",i)
        pb = run[i]["playbook"]
        targets = run[i]["targets"]
        cmd = "ansible-playbook " + pb
        for inv in inventory:
            cmd += " -i " + str(inv)

        target=""
        for t in targets:
            target += "," + t
        
        target = re.sub(r'(^,)','',target)

        cmd += " -l " + target
        print("cmd:",cmd)
        #if debug:
            #print("pb:", pb, "targets:", targets)

    #print("run main:", run)

File: roles/inventoryplay/files/inventory.pl
#!/usr/bin/perl -w

use strict;
use Data::Dumper;
use File::Basename;
use File::Temp qw/ tempdir /;
use Getopt::Long;
use Data::Dumper;


my $usage = "$0 -d|--directory <directory> -i|--inventory <ansible inventory> ... -i|--inventory=... --debug\n";
my $debug = 0;

#
# Get all servers that next playbook is $pb
# Incoming hash is $hash{$target}=\@arr (playbooks)
# Returns an array.
#

sub get_all_playbooks {
	my($pb) = shift;
	my(%hash) = @_;
	my(@res) = ();

	return(@res) unless ( $pb );
	print "searching for pb: $pb\n" if ( $debug );
	my($target);
	foreach $target ( sort keys %hash ) {
		my($arr) = $hash{$target};
		my($nextpb) = @$arr[0];
		if ( defined($nextpb) ) {
			if ( $nextpb eq $pb ) {
				push(@res,$target);
				shift(@$arr);
				$hash{$target} = $arr;
			}
		}
	}

	return(@res);
}

#
# Read all files in src ( one per target )
# The file contains the order of execution of playbooks.
#
# Example:
#   autopatch.yml
#   bepa.yml
#   site.yml
#   cepa.yml

#
# hash{target}@arr[playbook1, playbook2, playbookn]
#
sub read_all_files($) {
	my($dir) = shift;
	my(%hash);
	return(%hash) unless ( defined($dir) );
	if ( ! -d $dir ) {
		chdir($dir);
		print "chdir($dir): $!\n";
		return(%hash);
	}

	my($file);
	foreach $file ( <$dir/*> ) {
		my($target) = basename($file);
		my(@arr);
		print "Found: $target\n" if ( $debug );
		unless ( open(IN, "<", $file) ) {
			print "Reading $file: $!\n";
			next;
		}
		my($line);
		foreach $line ( <IN> ) {
			next unless ( defined($line) );
			next unless ( $line );
			chomp($line);
			push(@arr,$line);
		}
		close(IN);
		$hash{$target}=\@arr;
	}
	return(%hash);
}


#
# %hash{$run}{playbook}=playbook
# %hash{$run}{targets}=@arr(target,target,target...)
#
sub construct_runs(\%) {
	my($hash) = shift;
	my(%hash) = %$hash;

	my(%playbook);

	my($run) = 0;
	my($found) = 1;
	
	while ( $found ) {
		$found = 0;
		$run++;
	
		my($target);
		foreach $target ( sort keys %hash ) {
			print "Target: $target\n" if ( $debug );
			my($pb) = $hash{$target}[0];
			next unless ( $pb );
			$found++;
			if ( ! exists($playbook{$run}) ) {
				$playbook{$run}{playbook}=$pb;
				print "pb: $pb, run: $run\n" if ( $debug );
				#push(@playbook,$pb . ":" . $run);
				my(@arr) = get_all_playbooks($pb,%hash);
				$playbook{$run}{targets}=\@arr;
			}
		}
	}
	return(%playbook);
}

my $directory   = undef;
my @inventory   = undef;

GetOptions (
	"d|directory=s" => \$directory,
	"i|inventory=s" => \@inventory,
	"debug"  => \$debug,
  ) or die("Error in command line arguments\n");

unless ( defined($directory) and $#inventory > 0 ) {
	die $usage;
}

my ($dev,$ino,$mode,$nlink,$uid,$gid,$rdev,$size,
    $atime,$mtime,$ctime,$blksize,$blocks)
                       = stat($directory);

if ( ! defined($dev) ) {
	print $usage;
	die "Checking $directory $!\n";
}

my(%hash) = read_all_files($directory);

my(%playbook) = construct_runs(%hash);

my($run);
foreach $run ( sort { $a <=> $b } keys %playbook ) {
	print "run $run\n";
	my($playbook) = $playbook{$run}{playbook};
	my($targets) = $playbook{$run}{targets};
	my($subcmd) = "ansible-playbook $playbook";
        foreach ( @inventory ) {
		next unless ( $_ );
		$subcmd .= " -i $_";
	}
	$subcmd .= " -l " . join(",",@$targets);
	print $subcmd . "\n";
	my($rc) = 0;
	$rc = system($subcmd) unless ( $debug );
	print "rc: $rc  ($subcmd)\n\n\n";
}
